crates/client/contract-class-data/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+//
+// The buffer must have a capacity > 0. Use `new` to make sure it is conscructed correctly.
+pub struct DeclareTransactionJobReceiverFuture<B, P>
+where
+    P: TransactionPool<Block = B>,
+    B: BlockT,
+{
+    receiver: UnboundedReceiver<DeclareTransactionStatusStream<P>>,
+    buffer: Vec<DeclareTransactionStatusStream<P>>,
+}
+
+impl<B, P> DeclareTransactionJobReceiverFuture<B, P>
+where
+    P: TransactionPool<Block = B>,
+    B: BlockT,
+{
+    #[inline(always)]
+    pub fn new(receiver: UnboundedReceiver<DeclareTransactionStatusStream<P>>, limit: NonZeroUsize) -> Self {
+        Self { receiver, buffer: Vec::with_capacity(limit.into()) }
+    }
+
+    #[inline(always)]
+    pub fn buffer_as_mut_ref(&mut self) -> &mut Vec<DeclareTransactionStatusStream<P>> {
+        &mut self.buffer
+    }
+}
+
+// When calling `poll_next`, if new values have been pushed to the channel, stores up to limit of
+// them in the buffer, and returns the exact amount.
+//
+// This whole struct only makes sense if limit > 0.
+// We use `buffer.capacity()` as limit, so  use the `new` method
+// that won't let you pass a non allocated `Vec`.
+impl<B, P> Stream for DeclareTransactionJobReceiverFuture<B, P>
+where
+    P: TransactionPool<Block = B>,
+    B: BlockT,
+{
+    type Item = usize;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let DeclareTransactionJobReceiverFuture { receiver, buffer, .. } = &mut *self;
+
+        match receiver.poll_recv_many(cx, buffer, buffer.capacity()) {
+            Poll::Pending => Poll::Pending,
+            // For limit > 0, this means that the channel has been closed
+            Poll::Ready(0) => Poll::Ready(None),
+            Poll::Ready(amount) => Poll::Ready(Some(amount)),
+        }
+    }
+}
+
+pub async fn run_worker<B, P>(
+    contract_class_db: Arc<ContractClassDataDb>,
+    mapping_db: Arc<MappingDb<B>>,
+    job_recv: UnboundedReceiver<DeclareTransactionStatusStream<P>>,
+) where
+    B: BlockT,
+    P: TransactionPool<Block = B>,
+{
+    // We push everything we get from the receiver here, then repetitively iterate over it to poll
+    // each transaction's status stream and take action when they end up being finalized.
+    let mut unfinalized_declare_transactions = Vec::new();
+    // Stores all the txs that were not yet mapped into our db when they finalized for later processing.
+    let mut finalized_but_not_mapped_yet = HashMap::<B::Hash, Vec<(TransactionHash, ClassHash)>>::new();
+    // Totally arbitrary limit.
+    // The job_recv being unbounded, there is not risk of it being full.
+    // Could be increased up to at most the max amount of declare tx possible to have in a single block.
+    // More that this would be useless as the "poll receiver" task is woke up at least one (but most
+    // likely multiple times) during the span of one blocktime.
+    let streams_buffer_limit = unsafe { NonZeroUsize::new_unchecked(10) };
```

Comment:
No need, we will just take 10 more in a later iteration of the loop. And there is no risk of congestion because the channel is unbounded.

Even in the hypothesis of blocks full of DeclareTx, I think we will still be able to empty the channel easily.